File: 2328-number-of-increasing-paths-in-a-grid/2328-number-of-increasing-paths-in-a-grid.py
class Solution:
    def __init__(self):
        self.dp = []
        self.Mod = int(1e9) + 7
        self.dj = [0, 1, 0, -1]
        self.di = [-1, 0, 1, 0]
    
    def getPaths(self, grid, row, col):
        if self.dp[row][col]:
            return self.dp[row][col]
        
        ans = 0
        for d in range(4):
            i = row + self.di[d]
            j = col + self.dj[d]
            if 0 <= i < len(grid) and 0 <= j < len(grid[0]):
                if grid[i][j] > grid[row][col]:
                    ans = (ans + self.getPaths(grid, i, j)) % self.Mod
        
        ans = (ans + 1) % self.Mod
        self.dp[row][col] = ans
        return ans
    
    def countPaths(self, grid):
        m = len(grid)
        n = len(grid[0])
        self.dp = [[0] * n for _ in range(m)]
        
        ans = 0
        for i in range(m):
            for j in range(n):
                if self.dp[i][j]:
                    ans = (ans + self.dp[i][j]) % self.Mod
                else:
                    ans = (ans + self.getPaths(grid, i, j)) % self.Mod
        
        return ans

File: 2328-number-of-increasing-paths-in-a-grid/2328-number-of-increasing-paths-in-a-grid.cpp
class Solution {
public:
    vector<vector<long>> dp;
    int Mod = 1e9+7;
    vector<int> dj{0,1,0,-1};
    vector<int> di{-1,0,1,0};
    int getPaths(vector<vector<int>> &grid, int row, int col){
        if(dp[row][col]){
            return dp[row][col];
        }
        long long ans = 0;
        for(int d = 0; d < 4; d++){
            int i = row + di[d];
            int j = col + dj[d];
            if(i >= 0 && i < grid.size() && j >= 0 && j < grid[0].size()){
                if(grid[i][j] > grid[row][col]){
                    ans = (ans + getPaths(grid, i, j))%Mod;
                }
            }
        }
        ans = (ans+1)%Mod;
        return dp[row][col] = ans;
    }
    int countPaths(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        dp.resize(m, vector<long>(n,0));
        long long ans = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(dp[i][j]){
                    ans = (ans+dp[i][j])%Mod;
                }else{
                    ans = (ans + getPaths(grid, i, j))%Mod;
                }
            }
        }
        return ans;
    }
};